src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs

```diff
@@ -0,0 +1,177 @@
+        {
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var isApiController = controllerModel.Attributes.OfType<IApiBehaviorMetadata>().Any();
+                var controllerHasSelectorModel = controllerModel.Selectors.Any(s => s.AttributeRouteModel != null);
+
+                foreach (var actionModel in controllerModel.Actions)
+                {
+                    if (!isApiController && !actionModel.Attributes.OfType<IApiBehaviorMetadata>().Any())
+                    {
+                        continue;
+                    }
+
+                    if (!controllerHasSelectorModel && !actionModel.Selectors.Any(s => s.AttributeRouteModel != null))
+                    {
+                        // Require attribute routing with controllers annotated with ApiControllerAttribute
+                        throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+                    }
+
+                    if (_apiBehaviorOptions.EnableModelStateInvalidFilter)
+                    {
+                        Debug.Assert(_apiBehaviorOptions.InvalidModelStateResponseFactory != null);
+                        actionModel.Filters.Add(_modelStateInvalidFilter);
+                    }
+
+                    if (_modelMetadataProvider != null && _apiBehaviorOptions.InferBindingSourcesForParameters)
+                    {
+                        foreach (var parameter in actionModel.Parameters)
+                        {
+                            var bindingSource = InferBindingSourceForParameter(parameter);
+                            if (bindingSource != null)
+                            {
+                                _logger.InferringParameterBindingSource(parameter, bindingSource);
+                                parameter.BindingInfo = new BindingInfo
+                                {
+                                    BindingSource = bindingSource,
+                                };
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Internal for unit testing.
+        internal BindingSource InferBindingSourceForParameter(ParameterModel parameter)
+        {
+            if (parameter.BindingInfo != null)
+            {
+                return null;
+            }
+
+            if (ParameterExistsInAllRoutes(parameter.Action, parameter.ParameterName))
+            {
+                return BindingSource.Path;
+            }
+            else
+            {
+                var parameterMetadata = _modelMetadataProvider.GetMetadataForParameter(parameter.ParameterInfo);
+                if (parameterMetadata != null)
+                {
+                    var bindingSource = parameterMetadata.IsComplexType ?
+                        BindingSource.Body :
+                        BindingSource.Query;
```

Comment:
What is the experience like if you end up with multiple `[FromBody]` parameters? Can you easily figure out what's wrong?

Reply:
Would bailing at that point with Log.Info \ Log.Warn make sense?

Reply:
Let's roll this into a separate issue

Reply:
https://github.com/aspnet/Mvc/issues/6963